noir-projects/noir-protocol-circuits/crates/blob/src/blob.nr

```diff
@@ -265,6 +133,31 @@ pub fn evaluate_blob(
     BlobPublicInputs { z: challenge_z, y, kzg_commitment }
 }
 
+// Evaluates each blob required for a block
+pub fn evaluate_blobs(
+    blob_as_fields: [Field; FIELDS_PER_BLOB * BLOBS_PER_BLOCK],
+    kzg_commitments: [[Field; 2]; BLOBS_PER_BLOCK],
+    mut sponge_blob: SpongeBlob,
+) -> BlockBlobPublicInputs {
+    // Note that with multiple blobs per block, each blob uses the same hashed_blob_fields in:
+    // challenge_z = H(hashed_blob_fields, kzg_commitment[0], kzg_commitment[1])
+    // This is ok, because each commitment is unique to the blob, and we need hashed_blob_fields to encompass
+    // all fields in the blob, which it does.
+    let hashed_blob_fields = check_block_blob_sponge(blob_as_fields, sponge_blob);
+    let mut result = BlockBlobPublicInputs::empty();
+    for i in 0..BLOBS_PER_BLOCK {
+        let single_blob_fields = array_splice(blob_as_fields, i * FIELDS_PER_BLOB);
+        result.inner[i] = evaluate_blob(single_blob_fields, kzg_commitments[i], hashed_blob_fields);
+        if (result.inner[i].is_zero()) & (single_blob_fields[0] == 0) {
+            // We use empty PIs for empty blobs, to make it simpler to verify on L1.
+            // Since our fields come from the base rollup, we know they are tightly packed
+            // and should contain no 0 values among valid values => single_blob_fields[0] == 0.
+            result.inner[i] = BlobPublicInputs::empty();
+        }
+    }
+    result
+}
+
 /**
  *                    ___d-1
  *         z^d - 1    \            omega^i
```

Comment:
A nice optimisation (probably for a separate PR):

Given that we're handling 3 blobs in 1 circuit, we de-duplicate some of the computation of p(z):

<img width="378" alt="image" src="https://github.com/user-attachments/assets/1f9ef11a-ddf3-4f96-bdc3-3bb002f86226">

The only thing that differs with each blob (for a given block) is `y_i` (since they all share the same `z`). So we only need to compute the following once:
- `(z^d - 1) / d`
- `(w^i) / (z - w^i)`

That should speed things up nicely; especially getting rid of unnecessary inversions.

Reply:
Ooh nice! I will definitely work on this, maybe on another PR. Great spot!

Reply:
Just in case anyone comes back to this wondering why it wasn't done - each blob still has a different `z` for now, so we'll introduce the above if/when they are made the same